graph/formats/rdf/equi_canonical.go

```diff
@@ -0,0 +1,576 @@
+
+	var (
+		hasBlanks bool
+		err       error
+	)
+	for _, s := range g {
+		if isBlank(s.Subject.Value) || isBlank(s.Object.Value) {
+			hasBlanks = true
+			if err != nil {
+				break
+			}
+		}
+		if s.Label.Value != "" && err == nil {
+			err = errors.New("rdf: data-set contains graph names")
+			if hasBlanks {
+				break
+			}
+		}
+	}
+	if hasBlanks {
+		g = lean(&dfs{}, g)
+	}
+	return g, err
+}
+
+// removeRedundantBnodes removes blank nodes whose edges are a subset of
+// another term in the RDF graph.
+//
+// This is algorithm 4 in doi:10.1145/3068333.
+func removeRedundantBnodes(g []*Statement) []*Statement {
+	g = append(g[:0:0], g...)
+	for {
+		edges := make(map[string]map[triple]bool)
+		for _, s := range g {
+			for i, t := range []string{
+				s.Subject.Value,
+				s.Object.Value,
+			} {
+				e, ok := edges[t]
+				if !ok {
+					e = make(map[triple]bool)
+					edges[t] = e
+				}
+				switch i {
+				case 0:
+					e[triple{s.Predicate.Value, s.Object.Value, "+"}] = true
+				case 1:
+					e[triple{s.Predicate.Value, s.Subject.Value, "-"}] = true
+				}
+			}
+		}
+
+		seen := make(map[string]bool)
+		bNodes := make(map[string]bool)
+		terms := make(map[string]bool)
+		for _, s := range g {
+			for _, t := range []string{
+				s.Subject.Value,
+				s.Predicate.Value,
+				s.Object.Value,
+			} {
+				terms[t] = true
+				if isBlank(t) {
+					bNodes[t] = true
+				} else {
+					seen[t] = true
+				}
+			}
+		}
+
+		redundant := make(map[string]bool)
+		for x := range bNodes {
+			for xp := range terms {
+				if isProperSubset(edges[x], edges[xp]) || (seen[xp] && isEqualEdges(edges[x], edges[xp])) {
+					redundant[x] = true
+					break
+				}
+			}
+			seen[x] = true
+		}
+
+		n := len(g)
+		for i := 0; i < len(g); {
+			if !redundant[g[i].Subject.Value] && !redundant[g[i].Object.Value] {
+				i++
+				continue
+			}
+			g[i], g = g[len(g)-1], g[:len(g)-1]
+		}
+		if n == len(g) {
+			return g
+		}
+	}
+}
+
+type triple [3]string
+
+func isProperSubset(a, b map[triple]bool) bool {
+	for k := range a {
+		if !b[k] {
+			return false
+		}
+	}
+	return len(a) < len(b)
+}
+
+func isEqualEdges(a, b map[triple]bool) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for k := range a {
+		if !b[k] {
+			return false
+		}
+	}
+	return true
+}
+
+// findCandidates finds candidates for blank nodes and blank nodes that are fixed.
+//
+// This is algorithm 5 in doi:10.1145/3068333.
+func findCandidates(g []*Statement) ([]*Statement, map[string]bool, map[string]map[string]bool, bool) {
+	g = removeRedundantBnodes(g)
+
+	edges := make(map[triple]bool)
+	f := make(map[string]bool)
+	for _, s := range g {
+		sub := s.Subject.Value
+		prd := s.Predicate.Value
+		obj := s.Object.Value
+
+		edges[triple{sub, prd, obj}] = true
+		edges[triple{sub, prd, "*"}] = true
+		edges[triple{"*", prd, obj}] = true
+		switch {
+		case isBlank(sub) && isBlank(obj):
+			f[sub] = false
+			f[obj] = false
+		case isBlank(sub):
+			if _, ok := f[sub]; !ok {
```

Comment:
The translation here from the mathematical set notation to Go is fiendish.

Reply:
Yep.